### PR TITLE
streamproducer: retry loading a job in the c2c stream producer

### LINKED ISSUE #107989: c2c: long running cluster replication stopped (producer failed)
looks like the producer job hit:

```
load-job-query: replica unavailable: (n3,s3):2 unable to serve request to r17:/Table/1{5-6} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, (n5,s5):5, next=6, gen=8]: closed timestamp: 1690589503.985698795,0 (2023-07-29 00:11:43); raft status: {"id":"2","term":92,"vote":"1","commit":1567,"lead":"0","raftState":"StatePreCandidate","applied":1567,"progress":{},"leadtransferee":"0"}: operation "probe" timed out after 1m0.004s (given timeout 1m0s): result is ambiguous: after 60.00s of attempting command: context deadline exceeded
```

which is because `LoadJob` failed here https://github.com/cockroachdb/cockroach/blob/df7e2231cc3110da122bfe313044722c4818eac4/pkg/ccl/streamingccl/streamproducer/producer_job.go#L96C1-L96C1

we should retry that `LoadJob` call.

an additional fix can be to recreate the producer job from the destination (ingestion) job when it fails.

Jira issue: CRDB-30281

### PR DESCRIPTION
**Pull request by lidorcarmel**
See commits.

Epic: none
Fixes: #107989

Release note: None

### CODE REVIEW

File: pkg/ccl/streamingccl/streamproducer/producer_job.go [hunk 1]
```diff
@@ -93,9 +101,24 @@ func (p *producerJobResumer) Resume(ctx context.Context, execCtx interface{}) er
...
+				// No need to retry if there is no job record.
+				if err == nil || jobs.HasJobNotFoundError(err) {
+					break
+				}
+
+				log.Warningf(ctx, "replication stream %d failed, may retry: %v", p.job.ID(), err)
+			}
 			if err != nil {
-				return err
+				return errors.Wrapf(err, "replication stream %d cannot resume", p.job.ID())
```
Comment (stevendanna) [inline pkg/ccl/streamingccl/streamproducer/producer_job.go:121]: Lets mention that this came from loading the job

**Review by stevendanna — COMMENTED** [review]
This is certainly an improvement, but I wonder if we should go further.  Namely, perhaps we should retry for much longer and then moved to a paused state rather than a failed state.

Failing here is pretty drastic as it means that you potentially need to start a whole new tenant replication. 

Thoughts?

File: pkg/ccl/streamingccl/streamproducer/producer_job.go (hunk 1 above)
Comment (lidorcarmel) [reply pkg/ccl/streamingccl/streamproducer/producer_job.go]: done.

**Comment by lidorcarmel** [comment]
PTAL.
Thanks for the chat yesterday, this is still only the retry on the producer side, I'll have another PR for the ingestion side and another for changing the expiration to be last heartbeat as we discussed.

**Review by stevendanna — APPROVED** [review]

**Comment by stevendanna** [comment]
This looks good to me still.

**Comment by lidorcarmel** [comment]
hi @stevendanna! finally got back to this, you already reviewed the second commit, but the new test there was flaky because we didn't retry on the destination, so I added the first commit to do that retry. please review that 'retry heartbeating' commit.

**Review by stevendanna — APPROVED** [review]
Really nice testing here.

I think we discussed some strategies for putting an upper bound on this retrying by keeping the last-seen progress in memory and using that to drive some eventual expiration. But, let's keep that for a follow up.

I think we should backport this one.

**Review by lidorcarmel — COMMENTED** [review]
thanks Steven!
I think you're traveling so I'll merge, but if anything comes up later I'm happy to do a followup.

**Comment by lidorcarmel** [comment]
TFTR! and also thanks for the discussions Steven!
bors r+

**Pull request merged**